### PR TITLE
feat(engine-core): Prevent slotting in light DOM components

### PR DESCRIPTION
**Pull request by pmdartus**
## Details

With this PR, slotting content into a light DOM component throws an error in DEV mode. This restriction will be lifted once we enable support for slotting in the light DOM.

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅ 
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/vm.ts [hunk 1]
```diff
@@ -604,12 +604,6 @@ function getErrorBoundaryVM(vm: VM): VM | undefined {
 // NOTE: we should probably more this routine to the synthetic shadow folder
 // and get the allocation to be cached by in the elm instead of in the VM
 export function allocateInSlot(vm: VM, children: VNodes) {
-    if (process.env.NODE_ENV !== 'production') {
```
Comment (pmdartus) [inline packages/@lwc/engine-core/src/framework/vm.ts:607]: I was looking through the code, and this assertion doesn't make sense anymore as the `vm.cmpSlots` is always initialized.

File: packages/@lwc/engine-core/src/framework/hooks.ts [hunk 2]
```diff
@@ -161,8 +164,17 @@ export function allocateChildrenHook(vnode: VCustomElement, vm: VM) {
...
     const children = vnode.aChildren || vnode.children;
-
     vm.aChildren = children;
+
+    if (process.env.NODE_ENV !== 'production') {
+        assert.isTrue(
+            hasShadow(vm) || children.length === 0,
+            `Invalid usage of ${getComponentTag(
+                vm
+            )}. Light DOM components don't support yet slotting.`
```
Comment (ravijayaramappa) [inline packages/@lwc/engine-core/src/framework/hooks.ts:174]:
```suggestion
            )}. Light DOM components don't support slotting yet.`
```

File: packages/integration-karma/test/light-dom/slotting/index.spec.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+describe('slotted content', () => {
+    beforeEach(() => {
+        setFeatureFlagForTest('ENABLE_LIGHT_DOM_COMPONENTS', true);
+    });
+
+    afterEach(() => {
+        setFeatureFlagForTest('ENABLE_LIGHT_DOM_COMPONENTS', false);
+    });
+
+    it('should throws when attempting to render a slotted content in the light DOM', () => {
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/light-dom/slotting/index.spec.js:14]:
```suggestion
    it('should throw when attempting to render a slotted content in the light DOM', () => {
```

File: packages/integration-karma/test/light-dom/slotting/x/slotted/slotted.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { LightningElement } from 'lwc';
+
+export default class Slotted extends LightningElement {
+    static shadow = false;
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/light-dom/slotting/x/slotted/slotted.js:4]: IMO, this flag in the child does not matter.
```suggestion
```

**Review by ravijayaramappa — APPROVED** [review]
Minor typos otherwise LGTM

**Comment by ravijayaramappa** [comment]
IMO, similar restrictions should be added on the host element to prevent appending a child programmatically(in native shadow especially). That would go [here](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/restrictions.ts#L221). The engine currently has no restrictions to prevent adding child elements on the host.

**Review by nolanlawson — APPROVED** [review]

**Review by abdulsattar — APPROVED** [review]

**Review by ravijayaramappa — APPROVED** [review]

**Pull request merged**